core/src/main/java/kafka/server/SharePartition.java

```diff
@@ -596,6 +596,92 @@ public CompletableFuture<Optional<Throwable>> acknowledge(String memberId, List<
         return CompletableFuture.completedFuture(Optional.ofNullable(throwable));
     }
 
+    /**
+     * Release the acquired records for the share partition. The next fetch offset is updated to the next offset
+     * that should be fetched from the leader.
+     *
+     * @param memberId The member id of the client that is fetching/acknowledging the record.
+     *
+     * @return A future which is completed when the records are released.
+     */
+    public CompletableFuture<Optional<Throwable>> releaseAcquiredRecords(String memberId) {
+        log.trace("Release acquired records request for share partition: {}-{}", groupId, topicIdPartition);
+
+        Throwable throwable = null;
+        lock.writeLock().lock();
+        List<InFlightState> updatedStates = new ArrayList<>();
+        try {
+            long localNextFetchOffset = nextFetchOffset;
+            // TODO: recordState can be ARCHIVED as well, but that will involve maxDeliveryCount which is not implemented so far
+            RecordState recordState = RecordState.AVAILABLE;
+
+            // Iterate over multiple fetched batches. The state can vary per offset entry
+            for (Map.Entry<Long, InFlightBatch> entry : cachedState.entrySet()) {
+                InFlightBatch inFlightBatch = entry.getValue();
+
+                if (!inFlightBatch.memberId().equals(memberId)) {
+                    log.info("Member {} is not the owner of batch record {} for share partition: {}-{}",
+                            memberId, inFlightBatch, groupId, topicIdPartition);
+                    continue;
+                }
+
+                if (inFlightBatch.offsetState != null) {
+                    log.trace("Offset tracked batch record found, batch: {} for the share partition: {}-{}", inFlightBatch,
+                            groupId, topicIdPartition);
+                    for (Map.Entry<Long, InFlightState> offsetState : inFlightBatch.offsetState.entrySet()) {
+                        if (offsetState.getValue().state == RecordState.ACQUIRED) {
+                            InFlightState updateResult = offsetState.getValue().startStateTransition(recordState, false);
+                            if (updateResult == null) {
+                                log.debug("Unable to release records from acquired state for the offset: {} in batch: {}"
+                                                + " for the share partition: {}-{}", offsetState.getKey(),
+                                        inFlightBatch, groupId, topicIdPartition);
+                                throwable = new InvalidRecordStateException("Unable to release acquired records for the batch");
+                                break;
```

Comment:
I think this kind of situation indicates an inconsistency in the broker's data. Jumping out with half the records released and setting an exception doesn't really seem right. The log line is definitely a good idea. I think I'd tend to continue with the loop instead of breaking. It's a tough call though.

Reply:
If we encounter any situation where this error occurs then we rollback any changed state as well [here](https://github.com/confluentinc/kafka/pull/1145/files#diff-4ef2490e40ca0c63d51bf46caf866690472805e34d3752a919217b28f97045bcR670), so the inconsistency does not come into the picture. Either all acquired records are release or in case of any exception, none of them will be released.

Reply:
On second thoughts, if share persister works as expected, I don't think we are susceptible to getting an `InvalidRecordStateException`. So, we can get rid of exception handling completely. @apoorvmittal10 what do you think?

Reply:
Yes there will be a roll back if state cannot be transitioned. I am not sure on the question of share persister, as the write operations are locked on the state and once we see the lock available for moving the queue ahead by calling the share persister then it will move.